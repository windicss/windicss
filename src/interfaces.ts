import type { Utility } from './lib/utilities/handler';

import type { Property, Style, InlineAtRule, Keyframes } from './utils/style';

export type DictStr = { [key: string]: string };

export type DeepNestDictStr = { [key: string]: string | DeepNestDictStr };

export type NestObject = { [key: string]: string | string[] | NestObject };

export type DeepNestObject = { [key: string]: NestObject };

export type GenericNestObject<T> = { [key: string]: T | GenericNestObject<T> };

export type AnyObject = Record<string, unknown>;

export type AnyValue<T> = T;

export type Output = Property | Style | Style[] | undefined;

export type colorOpacity = {
  opacityVariable?: string
  opacityValue?: string
}
export type colorCallback = (colorOpacity: colorOpacity) => string;
export type colorObject = { [key: string]: string | colorCallback | colorObject };

export type FontSize =
  | string
  | [fontSize: string, letterSpacing?: string]
  | [
      fontSize?: string,
      options?: { letterSpacing?: string; lineHeight?: string }
    ];

export type DefaultFontSize = [
  fontSize: string,
  options: { letterSpacing?: string; lineHeight: string }
];

export type Layer = 'base' | 'utilities' | 'components'

export type ThemeUtil = (path: string, defaultValue?: unknown) => any;

export type ConfigUtil = (
  theme: ThemeUtil,
  {
    negative,
    breakpoints,
  }: {
    negative: (config: DictStr) => DictStr;
    breakpoints: (config: DictStr) => DictStr;
  }
) => { [key:string]: any };

export type PluginFunction = (utils: PluginUtils) => void;

export type PluginUtilOptions =
  | string[]
  | {
      variants?: string[];
      completions?: string[];
      respectPrefix?: boolean;
      respectImportant?: boolean;
      respectSelector?: boolean;
      layer?: Layer;
      order?: number;
      group?: string;
    };

export interface PluginBuilder {
  (handler: PluginFunction, config?: Config): PluginOutput;
  withOptions: <T = DictStr>(
    pluginFunction: (options: T) => PluginFunction,
    configFunction?: (options: T) => Config
  ) => PluginWithOptions<T>;
}

export interface PluginOutput {
  handler: PluginFunction;
  config?: Config;
  __isOptionsFunction?: false;
}

export interface PluginWithOptions<T = DictStr> {
  (options?: T): PluginOutputWithOptions<T>;
  __isOptionsFunction: true;
  __pluginFunction: (options: T) => PluginFunction;
  __configFunction: (options: T) => Config;
}

export interface PluginOutputWithOptions<T = DictStr> extends PluginOutput {
  __options: T;
}

export type ThemeType = ConfigUtil | Record<string, any> | undefined

export interface BaseTheme {
  vars: ThemeType
  orientation: ThemeType
  screens: ThemeType
  colors: ThemeType
  spacing: ThemeType
  animation: ThemeType
  backdropBlur: ThemeType
  backdropBrightness: ThemeType
  backdropContrast: ThemeType
  backdropGrayscale: ThemeType
  backdropHueRotate: ThemeType
  backdropInvert: ThemeType
  backdropOpacity: ThemeType
  backdropSaturate: ThemeType
  backdropSepia: ThemeType
  backgroundColor: ThemeType
  backgroundImage: ThemeType
  backgroundOpacity: ThemeType
  backgroundPosition: ThemeType
  backgroundSize: ThemeType
  blur: ThemeType
  borderColor: ThemeType
  borderOpacity: ThemeType
  borderRadius: ThemeType
  borderWidth: ThemeType
  boxShadow: ThemeType
  boxShadowColor: ThemeType
  brightness: ThemeType
  caretColor: ThemeType
  caretOpacity: ThemeType
  container: ThemeType
  contrast: ThemeType
  cursor: ThemeType
  divideColor: ThemeType
  divideOpacity: ThemeType
  divideWidth: ThemeType
  dropShadow: ThemeType
  fill: ThemeType
  flex: ThemeType
  flexGrow: ThemeType
  flexShrink: ThemeType
  fontFamily: ThemeType
  fontSize: ThemeType
  fontWeight: ThemeType
  gap: ThemeType
  gradientColorStops: ThemeType
  grayscale: ThemeType
  gridAutoColumns: ThemeType
  gridAutoRows: ThemeType
  gridColumn: ThemeType
  gridColumnEnd: ThemeType
  gridColumnStart: ThemeType
  gridRow: ThemeType
  gridRowStart: ThemeType
  gridRowEnd: ThemeType
  gridTemplateColumns: ThemeType
  gridTemplateRows: ThemeType
  height: ThemeType
  hueRotate: ThemeType
  inset: ThemeType
  invert: ThemeType
  keyframes: ThemeType
  letterSpacing: ThemeType
  lineHeight: ThemeType
  listStyleType: ThemeType
  margin: ThemeType
  maxHeight: ThemeType
  maxWidth: ThemeType
  minHeight: ThemeType
  minWidth: ThemeType
  objectPosition: ThemeType
  opacity: ThemeType
  order: ThemeType
  outline: ThemeType
  outlineColor: ThemeType
  perspective: ThemeType
  perspectiveOrigin: ThemeType
  padding: ThemeType
  placeholderColor: ThemeType
  placeholderOpacity: ThemeType
  ringColor: ThemeType
  ringOffsetColor: ThemeType
  ringOffsetWidth: ThemeType
  ringOpacity: ThemeType
  ringWidth: ThemeType
  rotate: ThemeType
  saturate: ThemeType
  scale: ThemeType
  sepia: ThemeType
  skew: ThemeType
  space: ThemeType
  stroke: ThemeType
  strokeWidth: ThemeType
  strokeDashArray: ThemeType
  strokeDashOffset: ThemeType
  tabSize: ThemeType
  textColor: ThemeType
  textOpacity: ThemeType
  textShadow: ThemeType
  textStrokeColor: ThemeType
  textStrokeWidth: ThemeType
  textDecorationColor: ThemeType
  textDecorationOpacity: ThemeType
  textDecorationLength: ThemeType
  textDecorationOffset: ThemeType
  textIndent: ThemeType
  transformOrigin: ThemeType
  transitionDuration: ThemeType
  transitionDelay: ThemeType
  transitionProperty: ThemeType
  transitionTimingFunction: ThemeType
  translate: ThemeType
  width: ThemeType
  zIndex: ThemeType

  // contributed by extensions
  aspectRatio: ThemeType
  filter: ThemeType
  backdropFilter: ThemeType
  lineClamp: ThemeType
  snapMargin: ThemeType
  snapPadding:ThemeType
  typography: ThemeType
}

export type Theme = Partial<BaseTheme> | ({ [ key:string ]: ThemeType } & { extend?: Theme & { extend?: undefined } })

export type Plugin =
  | PluginFunction
  | PluginOutput
  | PluginWithOptions<unknown>
  | PluginOutputWithOptions<unknown>;

export type Shortcut = string | NestObject;
// '@apply': 'font-bold hover:bg-red-500',
// 'background': 'white',

export type DarkModeConfig = 'class' | 'media' | false;

export type Handlers = {
  static?: boolean
  color?: boolean
  opacity?: boolean
  number?: boolean
  string?: boolean
  bracket?: boolean
  time?: boolean
  hex?: boolean
  nxl?: boolean
  fraction?: boolean
  size?: boolean
  variable?: boolean
  negative?: boolean
}

export type StyleArrayObject = { [key: string]: Style[] }
export type ResolvedVariants = { [key: string]: () => Style }
export type VariantTypes = 'screen' | 'theme' | 'state' | 'orientation'
export type AddPluginType = 'static' | 'utilities' | 'components' | 'preflights' | 'shortcuts'

export interface ExtractorResultDetailed {
  attributes?: {
    names: string[];
    values: string[];
  };
  classes?: string[];
  ids?: string[];
  tags?: string[];
  undetermined?: string[];
}

export interface Extractor {
  extractor: (content: string, id?: string) => ExtractorResultDetailed | Promise<ExtractorResultDetailed>,
  extensions: string[]
}

export interface ExtractOptions {
  /**
   * Globs of files to be included from extractions
   */
  include?: string[]
  /**
   * Globs of files to be excluded from extractions
   *
   * @default ['node_modules', '.git']
   */
  exclude?: string[]
  /**
   * Extractors to handle different file types.
   * Compatible with PurgeCSS
   */
  extractors?: Extractor[]
}

export interface PreflightOptions {
  /**
   * Include all the preflight regardless the template
   *
   * @default false
   */
  includeAll?: boolean

  /**
   * Safelist of preflight that will always be included in the generated CSS
   */
  safelist?: string | (string | string[])[];

  /**
   * Blocklist of preflight that will always be excluded in the generated CSS
   */
  blocklist?: string | (string | string[])[]

  /**
   * Alias for resolving preflight
   */
  alias?: Record<string, string>

  /**
   * @default true
   */
  includeBase?: boolean

  /**
   * @default true
   */
  includeGlobal?: boolean

  /**
   * @default true
   */
  includePlugin?: boolean
}

export interface Config {
  presets?: Config[];
  prefixer?: boolean;
  attributify?: boolean | {
    prefix?: string
    separator?: string
    disable?: string[]
  }
  separator?: string;
  important?: boolean | string;
  darkMode?: DarkModeConfig;
  theme?: Theme;
  variantOrder?: string[];
  plugins?: Plugin[];
  handlers?: Handlers;
  corePlugins?: (keyof BaseTheme)[] | string[] | { [ T in keyof BaseTheme ] : boolean } | { [ key:string ] : boolean };
  prefix?: string;
  exclude?: RegExp[];
  alias?: {[key:string]: string};
  shortcuts?: {[key:string]: Shortcut};

  // ===== Depreacted =====
  /**
   * @deprecated no longer needed for Windi CSS
   */
  purge?: unknown;
  /**
   * @deprecated no longer needed for Windi CSS
   */
  variants?: { [key: string]: string[] };
  /**
   * Fallback
   */
  [key:string]: any;
}

export interface FullConfig extends Config {
  /**
   * Safelist of utilities that will always be included in the generated CSS
   */
  safelist?: string | (string | string[])[];
  /**
   * Blocklist of utilities that will always be excluded in the generated CSS
   */
  blocklist?: string | (string | string[])[]
  /**
   * Extractions options
   */
  extract?: ExtractOptions
  /**
   * Preflight options
   * Set `false` to disable preflight
   *
   * @default true
   */
  preflight?: PreflightOptions | boolean
}

export interface DefaultTheme {
  colors: { [key: string]: string | { [key: string]: string } };
  container: { [key: string]: string | { [key: string]: string } };
  fontFamily: { [key: string]: string[] };
  fontSize: { [key: string]: DefaultFontSize };
  keyframes: { [key: string]: { [key: string]: string } };
  outline: { [key: string]: [outline: string, outlineOffset: string] };
}

export interface DefaultTheme {
  [key: string]:
    | DictStr
    | { [key: string]: string | { [key: string]: string } }
    | { [key: string]: string[] }
    | { [key: string]: DefaultFontSize };
}

export interface DefaultConfig {
  presets: string[];
  darkMode: DarkModeConfig;
  theme: DefaultTheme;
  variantOrder: string[];
  variants: { [key: string]: string[] };
  plugins: Plugin[];
}

export interface StaticUtility {
  [key: string]: {
    meta: { group: string, order: number },
    utility: { [key: string]: string | string[] };
  }
}

export interface DynamicUtility {
  [key: string]: (utility: Utility, { theme }: PluginUtils) => Output;
}

export interface PluginUtils {
  addDynamic: (
    key: string,
    generator: UtilityGenerator,
    options?: PluginUtilOptions
  ) => UtilityGenerator;
  addUtilities: (
    utilities: DeepNestObject,
    options?: PluginUtilOptions
  ) => Style[];
  addComponents: (
    components: DeepNestObject | DeepNestObject[],
    options?: PluginUtilOptions
  ) => Style[];
  addBase: (baseStyles: DeepNestObject) => Style[];
  addVariant: (name: string, generator: VariantGenerator) => Style | Style[];
  e: (selector: string) => string;
  prefix: (selector: string) => string;
  theme: (path: string, defaultValue?: unknown) => unknown;
  variants: (path: string, defaultValue?: string[]) => string[];
  config: (path: string, defaultValue?: unknown) => unknown;
}

export type VariantGenerator = (generator: {
  modifySelectors: (
    modifier: ({ className }: { className: string }) => string
  ) => Style;
  atRule: (name: string) => Style;
  pseudoClass: (name: string) => Style;
  pseudoElement: (name: string) => Style;
  parent: (name: string) => Style;
  child: (name: string) => Style;
  separator: string;
  style: Style;
}) => Style;

export type UtilityGenerator = (generator: {
  Utility: Utility;
  Style: {
    (selector: string, property?: Property | Property[] | undefined, important?: boolean): Style;
    generate: (
      parent?: string | undefined,
      property?: NestObject | undefined,
      root?: Style | undefined
    ) => Style[];
  };
  Keyframes: {
    (selector: string, property?: Property | Property[] | undefined, important?: boolean): Keyframes;
    generate: (name: string, children: {
        [key: string]: {
            [key: string]: string;
        };
    }) => Keyframes[];
  };
  Property: {
    (
      name: string | string[],
      value?: string | undefined,
      comment?: string | undefined,
      important?: boolean
    ): Property;
    parse: (
      css: string
    ) => Property | InlineAtRule | (Property | InlineAtRule)[] | undefined;
  };
}) => Output;

export interface Element {
  raw: string;
  start: number;
  end: number;
  variants: string[];
  content?: Element[] | string;
  func?: string;
  type: 'group' | 'func' | 'utility' | 'alias';
  important: boolean;
}

export interface Validata {
  className: string;
  raw: string;
  start: number;
  end: number;
  variants: string[];
  content?: string | Element[] | undefined;
  func?: string | undefined;
  type: 'group' | 'func' | 'utility' | 'alias';
  important: boolean;
  parent: Element | undefined;
}

export interface ProcessorCache {
  count: number; // plugin utilities count
  html: string[];
  attrs: string[];
  classes: string[];
  variants: string[];
  utilities: string[];
}

export interface PluginCache {
  core?: { [key:string]:boolean };
  static: StyleArrayObject; // utilities that don't need dynamically generated
  dynamic: { [key: string]: ((utility: Utility) => Output)};
  utilities: StyleArrayObject;
  components: StyleArrayObject;
  preflights: StyleArrayObject;
  shortcuts: StyleArrayObject;
  alias: { [key: string]: Element[] };
  completions: { [key: string]: string[] };
}

export interface VariantUtils {
  modifySelectors: (modifier: ({ className }: {
      className: string;
  }) => string) => Style;
  atRule: (name: string) => Style;
  pseudoClass: (name: string) => Style;
  pseudoElement: (name: string) => Style;
  parent: (name: string) => Style;
  child: (name: string) => Style;
}
