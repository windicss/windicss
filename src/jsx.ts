/* eslint-disable @typescript-eslint/no-empty-interface */
export type UtilityNames =
  | 'p'
  | 'm'
  | 'w'
  | 'h'
  | 'z'
  | 'border'
  | 'grid'
  | 'flex'
  | 'bg'
  | 'text'
  | 'font'
  | 'opacity'
  | 'animate'
  | 'transition'
  | 'transform'
  | 'align'
  | 'justify'
  | 'items'
  | 'block'
  | 'content'
  | 'pos'
  | 'box'
  | 'overflow'
  | 'underline'
  | 'list'
  | 'gradient'
  | 'divide'
  | 'gap'
  | 'ring'
  | 'icon'
  | 'container'
  | 'space'
  | 'table'
  | 'order'
  | 'place'
  | 'display'
  | 'shadow'
  | 'blend'
  | 'filter'
  | 'backdrop'
  | 'cursor'
  | 'outline'
  | 'select'

export type VariantNames =
  | 'hover'
  | 'active'
  | 'focus'
  | 'enabled'
  | 'dark'
  | 'light'
  | 'sm'
  | 'lg'
  | 'md'
  | 'xl'
  | 'xxl'
  | 'first'
  | 'last'
  | 'child'
  | 'root'
  | 'before'
  | 'after'
  | 'all'

export type AttributifyNames<Prefix extends string = ''> =
  | `${Prefix}${UtilityNames}`
  | `${Prefix}${VariantNames}`
  | `${Prefix}${VariantNames}:${UtilityNames}`

export interface AttributifyAttributes extends Partial<Record<AttributifyNames, string>> {}
