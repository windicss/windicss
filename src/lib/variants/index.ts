import { generateOrientations } from './orientation';
import { generateScreens } from './screen';
import { generateThemes } from './theme';
import { generateStates } from './state';

import type { Style } from '../../utils/style';
import type { BaseTheme, Config, DictStr } from '../../interfaces';

export type Variants = {
  orientation: { [key: string]: () => Style };
  screen: { [key: string]: () => Style };
  theme: { [key: string]: () => Style };
  state: { [key: string]: () => Style };
}

export function resolveVariants(config: Config): Variants {
  return {
    orientation: generateOrientations(((config.theme as BaseTheme)?.orientation ?? {})),
    screen: generateScreens(((config.theme as BaseTheme)?.screens ?? {}) as DictStr),
    theme: generateThemes(config.darkMode),
    state: generateStates(config.variantOrder ?? []),
  };
}

export {
  generateOrientations,
  generateScreens,
  generateThemes,
  generateStates,
};
