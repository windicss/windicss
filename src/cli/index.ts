import arg from 'arg';
import { deepCopy, Console } from '../utils/tools';
import { resolve, dirname, join, extname } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFile, watch, unwatchFile, existsSync } from 'fs';
import { HTMLParser, CSSParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  globArray,
  generateTemplate,
  fuzzy,
} from './utils';
import type { Extractor } from '../interfaces';

const doc = `Generate css from text files that containing windi classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m -d
  windicss [filenames] -c -s -m -d
  windicss [filenames] [-c | -i] [-a] [-b | -s] [-m] [-d] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class selectors. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -a, --attributify     Attributify mode, generate attribute selectors. Attributify mode can be mixed with the other two modes.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -d, --dev             Enable hot reload and watch mode.
  -m, --minify          Generate minimized css file.
  -z, --fuzzy           Enable fuzzy match, only works in interpration mode.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --style               Parse and transform windi style block.
  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--attributify': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--dev': Boolean,
  '--minify': Boolean,
  '--fuzzy': Boolean,
  '--style': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-a': '--attributify',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',
  '-d': '--dev',
  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
  '-z': '--fuzzy',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const configFile = args['--config'] ? resolve(args['--config']) : undefined;
let preflights: { [key:string]: StyleSheet } = {};
let styleSheets: { [key:string]: StyleSheet } = {};
let processor = new Processor(configFile ? require(configFile) : undefined);
let safelist = processor.config('safelist');

if (configFile) Console.log('Config file:', configFile);

function compile(files: string[]) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  files.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true, args['--dev']); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    const added = (
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );
    styleSheets[file] = args['--dev'] ? (styleSheets[file]? styleSheets[file].extend(added) : added) : added;

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFile(outputFile, outputHTML.join(''), () => null);
    Console.log(`${file} -> ${outputFile}`);
    if (args['--preflight']) {
      if (args['--dev']) {
        const preflight = processor.preflight(html, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      } else {
        preflights[file] = processor.preflight(html);
      }
    }
  });
}

function interpret(files: string[]) {
  // interpretation mode
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    let classes: string[] = [];
    if (args['--fuzzy']) {
      classes = fuzzy(content);
    } else {
      const parser = new HTMLParser(content);
      classes = parser.parseClasses().map((i) => i.result);
    }
    const extractors = processor.config('extract.extractors') as Extractor[] | undefined;
    if (extractors) {
      for (const { extractor, extensions } of extractors) {
        if (extensions.includes(extname(file).slice(1))) {
          const result = extractor(content);
          if ('classes' in result && result.classes) {
            classes = [...classes, ...result.classes];
          }
        }
      }
    }
    if (args['--dev']) {
      const utility = processor.interpret(classes.join(' '), true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
      if (args['--preflight']) {
        const preflight = processor.preflight(content, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      }
    } else {
      const utility = processor.interpret(classes.join(' '));
      styleSheets[file] = utility.styleSheet;
      if (args['--preflight']) preflights[file] = processor.preflight(content);
    }
  });
}

function attributify(files: string[]) {
  // attributify mode
  files.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const attrs: { [key: string]: string | string[] } = parser
      .parseAttrs()
      .reduceRight((a: { [key: string]: string | string[] }, b) => {
        if (b.key === 'class' || b.key === 'className') return a;
        if (b.key in a) {
          a[b.key] = Array.isArray(a[b.key])
            ? Array.isArray(b.value)? [ ...(a[b.key] as string[]), ...b.value ]: [ ...(a[b.key] as string[]), b.value ]
            : [ a[b.key] as string, ...(Array.isArray(b.value) ? b.value : [b.value]) ];
          return a;
        }
        return Object.assign(a, { [b.key]: b.value });
      }, {});
    if (args['--dev']) {
      const utility = processor.attributify(attrs, true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
    } else {
      const utility = processor.attributify(attrs);
      styleSheets[file] = utility.styleSheet;
    }
  });
}

function styleBlock(files: string[]) {
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    const block = content.match(/(?<=<style lang=['"]windi["']>)[\s\S]*(?=<\/style>)/);
    if (block && block.index) {
      const css = content.slice(block.index, block.index + block[0].length);
      const parser = new CSSParser(css, processor);
      styleSheets[file] = styleSheets[file].extend(parser.parse());
    }
  });
}

function build(files: string[], update = false) {
  if (args['--compile']) {
    compile(files);
  } else {
    interpret(files);
  }
  if (args['--attributify']) attributify(files);
  if (args['--style']) styleBlock(files);
  if (args['--separate']) {
    for (const [file, sheet] of Object.entries(styleSheets)) {
      const outfile = file.replace(/\.\w+$/, '.windi.css');
      writeFile(outfile, (args['--preflight'] ? deepCopy(sheet).extend(preflights[file], false) : sheet).build(args['--minify']), () => null);
      Console.log(`${file} -> ${outfile}`);
    }
  } else {
    let outputStyle = Object.values(styleSheets)
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .sort()
      .combine();
    if (args['--preflight'])
      outputStyle = Object.values(preflights)
        .reduce(
          (previousValue: StyleSheet, currentValue: StyleSheet) =>
            previousValue.extend(currentValue),
          new StyleSheet()
        )
        .sort()
        .combine()
        .extend(outputStyle);
    const filePath = args['--output'] ?? 'windi.css';
    writeFile(filePath, outputStyle.build(args['--minify']), () => null);
    if (!update) {
      Console.log('Matched files:', files);
      Console.log('Output file:', resolve(filePath));
    }
  }

}

function buildSafeList(safelist: unknown) {
  if (safelist) {
    let classes: string[] = [];
    if (typeof safelist === 'string') {
      classes = safelist.split(/\s/).filter(i => i);
    }
    if (Array.isArray(safelist)) {
      for (const item of safelist) {
        if (typeof item === 'string') {
          classes.push(item);
        } else if (Array.isArray(item)) {
          classes = classes.concat(item);
        }
      }
    }
    styleSheets['safelist'] = processor.interpret(classes.join(' ')).styleSheet;
  }
}

const patterns = args._
  .concat(processor.config('extract.include', []) as string[])
  .concat((processor.config('extract.exclude', []) as string[]).map(i => '!' + i));

let matchFiles = globArray(patterns);

if (matchFiles.length === 0) {
  Console.error('No files were matched!');
  process.exit();
}

buildSafeList(safelist);
build(matchFiles);

function watchBuild(file: string) {
  watch(file, (event, path) => {
    if (event === 'rename') {
      const newFiles = globArray(patterns);
      const renamed = matchFiles.filter(i => !(newFiles.includes(i)))[0];
      if (existsSync(path)) {
        Console.log('File', `'${renamed}'`, 'has been renamed to', `'${path}'`);
        matchFiles = newFiles;
        Console.log('Matched files:', matchFiles);
      } else {
        Console.log('File', `'${file}'`, 'has been deleted');
        unwatchFile(file);
        matchFiles = newFiles;
        delete styleSheets[file];
        delete preflights[file];
        if (matchFiles.length > 0) {
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
        }
        build([], true);
        if (matchFiles.length > 0) {
          Console.timeEnd('Building');
        } else {
          Console.error('No files were matched!');
          process.exit();
        }
      }
    }
    if (event === 'change') {
      Console.log('File', `'${path}'`, 'has been changed');
      Console.time('Building');
      build([file], true);
      Console.timeEnd('Building');
    }
  });
}

function watchConfig(file?: string) {
  if (!file) return;
  let stamp = 0;
  watch(file, (event, path) => {
    if (event === 'change' && (stamp === 0 || + new Date() - stamp > 500)) {
      // fix fire twice event when change config file
      stamp = + new Date();
      Console.log('Config', `'${path}'`, 'has been changed');
      Console.time('Building');
      configFile && delete require.cache[configFile];
      processor = new Processor(configFile ? require(configFile) : undefined);
      safelist = processor.config('safelist');
      styleSheets = {};
      preflights = {};
      buildSafeList(safelist);
      build(matchFiles, true);
      Console.timeEnd('Building');
    }
  });
}

if (args['--dev']) {
  watchConfig(configFile);
  for (const file of matchFiles) {
    watchBuild(file);
  }
  for (const dir of Array.from(new Set(matchFiles.map(f => dirname(f))))) {
    watch(dir, (event, path) => {
      if (event === 'rename' && existsSync(join(dir, path))) {
        // when create new file
        const newFiles = globArray(patterns);
        if (newFiles.length > matchFiles.length) {
          const newFile = newFiles.filter(i => !matchFiles.includes(i))[0];
          Console.log('New file', `'${newFile}'`,  'added');
          matchFiles.push(newFile);
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
          build([newFile], true);
          watchBuild(newFile);
          Console.timeEnd('Building');
        }
      }
    });
  }
}
