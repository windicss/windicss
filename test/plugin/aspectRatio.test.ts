import { Processor } from '../../src/lib';
import aspectRatio from '../../src/plugin/aspect-ratio';

describe('aspect ratio plugin', () => {
  it('interpret test', () => {
    const processor = new Processor();
    processor.loadPlugin(aspectRatio);
    const classes = `
      aspect-none
      aspect-auto
      aspect-square
      aspect-video
      aspect-w-16
      aspect-h-9
      aspect-9/16
      `;
    const utility = processor.interpret(classes);
    expect(utility.ignored.length).toEqual(0);
    expect(utility.styleSheet.build()).toMatchSnapshot('aspect-ratio base');
  });

  it('works with prefix', () => {
    const processor = new Processor({ prefix: 'windi-' });
    processor.loadPlugin(aspectRatio);
    const classes = `
      windi-aspect-none
      windi-aspect-auto
      windi-aspect-square
      windi-aspect-video
      windi-aspect-w-16
      windi-aspect-h-9
      sm:windi-aspect-9/16
      `;
    const utility = processor.interpret(classes);
    expect(utility.ignored.length).toEqual(0);
    expect(utility.styleSheet.build()).toMatchSnapshot('aspect-ratio with prefix');
  });

  it('add completions', () => {
    const processor = new Processor();
    processor.loadPlugin(aspectRatio);
    expect(processor._plugin.completions).toMatchSnapshot('completions');
  });
});
