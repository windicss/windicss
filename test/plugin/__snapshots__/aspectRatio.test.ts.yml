Tools / add completions / completions / 0: |-
  {
    "aspectRatio": [
      "aspect-w-{float}",
      "aspect-h-{float}",
      "aspect-{fraction}"
    ]
  }
Tools / interpret test / aspect-ratio base / 0: |-
  .aspect-none {
    position: static;
    padding-bottom: 0;
  }
  .aspect-none > * {
    position: static;
    height: auto;
    width: auto;
    top: auto;
    right: auto;
    bottom: auto;
    left: auto;
  }
  .aspect-auto {
    aspect-ratio: auto;
  }
  .aspect-square {
    aspect-ratio: 1 / 1;
  }
  .aspect-video {
    aspect-ratio: 16 / 9;
  }
  .aspect-w-16 {
    --tw-aspect-w: 16;
    position: relative;
    padding-bottom: calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%);
  }
  .aspect-w-16 > * {
    position: absolute;
    height: 100%;
    width: 100%;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
  .aspect-h-9 {
    --tw-aspect-h: 9;
  }
  .aspect-9\/16 {
    position: relative;
    padding-bottom: 56.25%;
  }
  .aspect-9\/16 > * {
    position: absolute;
    height: 100%;
    width: 100%;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
Tools / works with prefix / aspect-ratio with prefix / 0: |-
  .windi-aspect-none {
    position: static;
    padding-bottom: 0;
  }
  .windi-aspect-none > * {
    position: static;
    height: auto;
    width: auto;
    top: auto;
    right: auto;
    bottom: auto;
    left: auto;
  }
  .windi-aspect-auto {
    aspect-ratio: auto;
  }
  .windi-aspect-square {
    aspect-ratio: 1 / 1;
  }
  .windi-aspect-video {
    aspect-ratio: 16 / 9;
  }
  .windi-aspect-w-16 {
    --tw-aspect-w: 16;
    position: relative;
    padding-bottom: calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%);
  }
  .windi-aspect-w-16 > * {
    position: absolute;
    height: 100%;
    width: 100%;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
  .windi-aspect-h-9 {
    --tw-aspect-h: 9;
  }
  @media (min-width: 640px) {
    .sm\:windi-aspect-9\/16 {
      position: relative;
      padding-bottom: 56.25%;
    }
    .sm\:windi-aspect-9\/16 > * {
      position: absolute;
      height: 100%;
      width: 100%;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
    }
  }
